### DISCUSSION TITLE
How to listen for account change? (Solved)

### DISCUSSION THREAD

**Discussion by Ian-Bright** [wevm/wagmi#1806] [Q&A]
Is there a hook where we can detect an account change? Specifically I'd like to force a disconnect if I user switches accounts on metamask while connected on one account

--EDIT: Nevermind found in a prior discussion

**Comment by allenchuang** [comment]
FYI You can use the following:

```
  const { connector: activeConnector } = useAccount()

  useEffect(() => {
	const handleConnectorUpdate = ({account, chain}: ConnectorData) => {
      if (account) {
        console.log('new account', account)
      } else if (chain) {
        console.log('new chain', chain)
      }
    }

    if (activeConnector) {
      activeConnector.on('change', handleConnectorUpdate)
    }

	return () => activeConnector.off('change', handleConnectorUpdate)
  }, [activeConnector])
  ```

**Reply by Melancholism** [reply]
Hi. Thanks for your response. I have an issue. When using your snippet, I face these errors: 
1)
> Argument of type '() => () => Connector<any, any>' is not assignable to parameter of type 'EffectCallback'.
  Type '() => Connector<any, any>' is not assignable to type 'void | Destructor'.
    Type '() => Connector<any, any>' is not assignable to type 'Destructor'.
      Type 'Connector<any, any>' is not assignable to type 'void | { [UNDEFINED_VOID_ONLY]: never; }'.ts(2345) 

2)

> 'activeConnector' is possibly 'undefined'

I managed to solve them this way:
```js
const { connector: activeConnector } = useAccount();
	React.useEffect(() => {
		const handleConnectorUpdate = ({ account, chain }: ConnectorData) => {
			if (account) {
				console.log("new account", account);
			} else if (chain) {
				console.log("new chain", chain);
			}
		};

		if (activeConnector) {
			activeConnector.on("change", handleConnectorUpdate);
		}

		return () => activeConnector?.off("change", handleConnectorUpdate) as any;
	}, [activeConnector]);
```

Everything works fine but I wanted to double-check if there is any downside to this workaround. I'd be appreciated to get your feedback.

**Reply by cosmicknox** [reply]
thanks alot works perfect :) added it to my  wallet context provider

**Reply by Melancholism** [reply]
@cosmicknox Glad I could help:)

**Reply by Conviley** [reply]
The problem is that `useEffect` cleanup function is returning `activeConnector?.off("change", handleConnectorUpdate)` which is of type `Connector<any,any>` whereas `useEffect` expects a cleanup function that returns nothing i.e `void`.

Rather than typecasting with `any` effectively disabling typescript may i suggest simply opening up the cleanup functions with brackets and only running but not returning the `off` function. i.e

```typescript
return () => {
  activeConnector?.off("change", handleConnectorUpdate);
};
```

**Reply by Emkay860** [reply]
This solution works, Thanks.

**Reply by raffle-x-dev** [reply]
What's ConnectorData? My typescript can't find it.

**Reply by ybgbob** [reply]
It look likes v1's solution

**Comment by mwawrusch** [comment]
> 

If you happen to find a solution please also post if for everyone else

**Comment by VitorPChaves** [comment]
Hey, can you share the previous discussion you mentioned on the edit? Can't find it.

**Comment by JorisZierold** [comment]
An even simpler solution:

```
React.useEffect(() => {
        const unwatch = watchAccount(account => {
            console.log('Account changed to:', account);
        });

        // Cleanup by calling unwatch to unsubscribe from the account change event
        return () => unwatch();
    }, []);
```

As of: https://wagmi.sh/core/actions/watchAccount

**Reply by Nilbristi** [reply]
In this function we can only see the console, but you can't assign that data or call any other function here. What is the use of this function then?

**Reply by Nilbristi** [reply]
We can only console data here and nothing else. Even can't assign that data to a variable or call any function. Then what is the use of this function?
```
const unwatch = watchAccount(config, {
      onChange(data) {
        console.log('Account changed!', data);
        <I want to call some function here or pass this data or update some variable>
      }
    });
```

**Reply by thatguyintech** [reply]
You should be able to get access to the current connected account and the previously connected account via the input args:

```
onChange(account: GetAccountReturnType, prevAccount: GetAccountReturnType): void`
```

```
  const config = useConfig();
  useEffect(() => {
    const unwatch = watchAccount(config, {
      onChange: (
        account: GetAccountReturnType,
        prevAccount: GetAccountReturnType,
      ) => {
        if (prevAccount.address !== account.address) {
          disconnect();
        }
      },
    });
    return () => unwatch();
  }, []);
  ```

**Reply by sbeben** [reply]
Argument of type 'Config<PublicClient, WebSocketPublicClient>' is not assignable to parameter of type 'WatchAccountCallback<PublicClient>'.

**Reply by sbeben** [reply]
and obviously this does not work

**Reply by ghost** [reply]
can not detect switching metamask account

**Reply by dimisus** [reply]
Nope, this is not the event it promises to be. This will not catch the switching accounts inside the wallet like MetaMask.

**Reply by dimisus** [reply]
> Nope, this is not the event it promises to be. This will not catch the switching accounts inside the wallet like MetaMask.

Correction, it will, but only for the accounts that have previously granted permissions!

**Comment by 1antares1** [comment]
For now, these workarounds aren't possible for me now or are deprecated...

But using the `useAccountEffect` available from wagmi, it has worked for me so far (it recognizes when the user disconnects inside and outside the app from their wallet): 👍🏻
E.g.:

```typescript
useAccountEffect({
    onConnect(data) {
      console.log('Connected!', data)
    },
    onDisconnect() {
      console.log('Disconnected!')
    },
  })
```

Ref link: https://wagmi.sh/react/api/hooks/useAccountEffect#useaccounteffect

**Reply by c0rdeiro** [reply]
but if the user changes accounts on metamask directly without disconnecting, does it trigger the onConnect? 
I had this issue and it wasn't entering the onConnect when the user changed accounts, and tecnically connecting via metamask

**Reply by Nilbristi** [reply]
NO, For that you need to set useAccountEffect
